### PR TITLE
Submit delegation certificates

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#893 

# Overview

- [x] I continued on Rodney's #947 and added two passing golden tests, which means `mkCertificateTx` should be working.

# Comments

- I am seriously considering doing more drastic changes regarding the responsibility of `TransactionLayer` and `C.W.J.Binary`.

### CODE REVIEW

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 1]
```diff
@@ -407,22 +409,35 @@ putStakeCertificate
     -> ChimericAccount
     -> Put
 putStakeCertificate (PoolId poolId) (ChimericAccount accId) = do
+    assertLength "poolId" poolId 32
+    assertLength "account pubKey" accId 32
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:412]: It's actually easy to mess up here, if you don't remember if the keys should be in hex or raw binary when constructing the newtypes.

Only exposing explicitly named smart constructors could be an idea.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api.hs [hunk 2]
```diff
@@ -186,11 +184,6 @@ instance MimeRender JormungandrBinary (Tx, [TxWitness]) where
         runPut $ withHeader MsgTypeTransaction $
             putSignedTx mempty ins outs wits
 
-instance MimeRender JormungandrBinary (ByteString, (Tx, [TxWitness])) where
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api.hs:189]: We would need `(payload, tx, witnesses, auth)` 😩 

I simply removed it here and used a single `ByteString` as return type of `mkCertificateTx`. Maybe we could use the `Fragment` type. Or some new `GenericTx` type.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 3]
```diff
@@ -500,8 +514,8 @@ putSignedTx payload inputs outputs witnesses = do
     putWitness :: TxWitness -> Put
     putWitness (TxWitness bytes) = putByteString bytes
 
-putTx :: [(TxIn, Coin)] -> [TxOut] -> Put
-putTx inputs outputs = do
+putTx :: Put -> [(TxIn, Coin)] -> [TxOut] -> Put
+putTx putPayload inputs outputs = do
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:517]: `putTx` is called by both `signData` and `putSignedTx` witch both should `putPayload`. So this is a minor refactoring.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 4]
```diff
@@ -407,22 +409,35 @@ putStakeCertificate
...
-putStakeDelegationTx poolId accId (Hash accSig) inputs outputs witnesses = do
-    putStakeCertificate poolId accId
-    putSignedTx mempty inputs outputs witnesses
-    putByteString accSig
+putStakeDelegationTx payload inputs outputs witnesses auth = do
+    withHeader MsgTypeDelegation $ do
+        withAppendedSignature auth $ do
+            putSignedTx payload inputs outputs witnesses
+
+withAppendedSignature :: (ByteString -> ByteString) -> Put -> Put
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:434]: why not to put it in `where` of `putStakeDelegationTx` if we do not use it anywhere else?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 5]
```diff
@@ -133,9 +148,11 @@ newTransactionLayer (Hash block0H) = TransactionLayer
         wits <- forM rnps $ \(_, TxOut addr _) -> do
             xprv <- maybeToRight (ErrKeyNotFoundForAddress addr) (keyFrom addr)
             let body = block0H <> getHash (signData payload inps outs)
-            pure $ mkTxWitness (fst xprv) (sign body xprv)
+            pure $ mkTxWitness (fst xprv) (sign xprv body)
         let tx = Tx
-                { txId = fragmentId msgType payload inps outs wits
+                { txId = error "we shouldn't try to calculate this yet"
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: I bet this is temporary and to be corrected in coming commits

File: lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs [hunk 6]
```diff
@@ -141,29 +152,36 @@ spec = do
...
-                        $ withHeader MsgTypeDelegation
-                        $ putStakeDelegationTx poolId accId accSig inps outs wits
-                let decode =
-                        getStakeDelegationTxMessage . runGet getMessage
-                tx' <- try' (decode encode)
-                if tx' == Right (poolId, accId, tx, wits)
-                then return ()
-                else expectationFailure $
-                    "tx /= decode (encode tx) == " ++ show tx'
+--         it "decode (encode tx) === tx stake delegation transaction" $
```
Comment (paweljakubas) [inline lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs:155]: why it is commented?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs (hunk 5 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: Yes – this is what makes me believe that we have to do something radically different.

I'm looking at completely diverging from this PR, and doing some refactoring first.

File: lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs [hunk 7]
```diff
@@ -234,11 +262,29 @@ instance Arbitrary StakeDelegationTx where
...
+
+instance {-# OVERLAPS #-} Arbitrary (ShelleyKey 'AddressK XPrv, Passphrase "encryption")
+  where
+    shrink _ = []
+    arbitrary = do
+        seed <- Passphrase . BA.convert . BS.pack <$> replicateM 32 arbitrary
+        pwd <- arbitrary
+        let root = generateKeyFromSeed (seed, mempty) pwd
+        let acc = deriveAccountPrivateKey pwd root minBound
+        let key = deriveAddressPrivateKey pwd acc MutableAccount minBound
```
Comment (paweljakubas) [inline lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs:280]: :+1:

File: lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs [hunk 8]
```diff
@@ -219,60 +219,57 @@ mkStdTxSpec = do
...
+        --     | jcli transaction finalize \
+        --     | jcli transaction seal \
+        --     | jcli transaction auth -k key.prv \
+        --     | jcli transaction to-message
+        let poolId = PoolId $ unsafeFromHex "5aac0894709438314ec1c8b0697820fc2efaac914768b53e54d68b0bce17a9a8"
+        goldenTestCertificateTx tl keystore
+            (poolId, xprv0, Passphrase "")
+            [] -- no inputs; we assume 0 fees here for simplicity
+            [] -- no outputs; no change
+            "008504002d155d07184387ba6ff8836014ba0b626b1808e67e75dfa370c2dc5f581715fe015aac0894709438314ec1c8b0697820fc2efaac914768b53e54d68b0bce17a9a8000005458a17be5c072b8981f43a272b1d8136f880ff91ff87d1b34aef3e860300df40997b2130b7db70796b30684ea99d5dd7de1767bd6602da7138864e30a3d202"
```
Comment (paweljakubas) [inline lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:250]: why not comply with 80-character line standard, here?

File: lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs [hunk 9]
```diff
@@ -219,60 +219,57 @@ mkStdTxSpec = do
...
-            [ (TxIn txin0 0, TxOut addr1 (Coin 10000))
-            ]
-            [ TxOut addr1 (Coin 9999)
-            ]
-            "00d604b2ddcfbf631000216a6e0410e64c20e6af91bbb2826276ad6bfda777e766b24bc7676b9c29cd3b97cc8cb297c44314f7bfee81d96661278765a82e9a91abf8710101000000000000002710666984dec4bc0ff1888be97bfe0694a96b35c58d025405ead51d5cc72a3019f403b2ddcfbf631000216a6e0410e64c20e6af91bbb2826276ad6bfda777e766b24b000000000000270f01ac2efe363b1c001497f04bc8bff72dacdc0f15bef295d2aae0dbb7b5e3371b95c5c571ec1f6ecf452d6ccc72fe0c4d8e2a1d155b9e02ebcedab530aa5b547f05"
+            (poolId, xprv0, pwd0)
+            [(TxIn txin0 0, TxOut addr0 (Coin 42))]
+            [] -- no outputs; no change
+            "00ef04002d155d07184387ba6ff8836014ba0b626b1808e67e75dfa370c2dc5f581715fe015aac0894709438314ec1c8b0697820fc2efaac914768b53e54d68b0bce17a9a8010000000000000000002a666984dec4bc0ff1888be97bfe0694a96b35c58d025405ead51d5cc72a3019f40179e6b8dc3885f26bea38de7e48adeb985e1d70b79aa1f6e150c2b61795dbb8d2e2fb76ae59ca862f28431ab318ef622999c25a4ed4947f299b5020a74317e70e62aee0d27ff232fa09b588483ba861fc374df7900836e07eb80bd2fc88684eb32d9b500db025cd0c3f8b83a86c6ac61d9426e61e9999a80e7ea19df8fe49c701"
```
Comment (paweljakubas) [inline lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:272]: same here

**Review by paweljakubas — APPROVED** [review]
Almost there (I guess small refactorings left). I am really fond of `use-Put-until-needed` style

**Comment by Anviking** [comment]
Closing this in favour of #1059. Can always be re-opened if #1021 does not go well.

**Pull request closed without merging**